File: 백준/Gold/1005. ACM Craft/ACM Craft.py
import sys
from collections import deque
input = sys.stdin.readline

T = int(input())

for _ in range(T):
    N, K = map(int, input().split())
    graph = [[] for _ in range(N+1)]
    time = [0] +  list(map(int, input().split()))
    # 위상 정렬, 진입차수 개념이 중요.
    inDegree = [0 for _ in range(N+1)]
    dp = [0 for _ in range(N+1)]

    for _ in range(K):
        a, b = map(int, input().split())
        graph[a].append(b)
        # 경로가 생길 때마다 +1
        inDegree[b] += 1

    W = int(input())
    queue = deque()

    for i in range(1, N+1):
        if inDegree[i] == 0:
            queue.append(i)
            # 처음 만들어 질 수 있는 시간은 첫  time idx
            dp[i] = time[i]
    
    while queue:
        build = queue.popleft()
        for i in graph[build]:
            inDegree[i] -= 1
            dp[i] = max(dp[i], dp[build] + time[i])
            if inDegree[i] == 0:
                queue.append(i)
    
    print(dp[W])

File: 백준/Gold/1005. ACM Craft/ACM Craft.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;
import java.util.Queue;
import java.util.LinkedList;

public class Main {
  static int t, n, k, w;
  static int[] time, inDegree, dp;
  static List<Integer>[] graph;
  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringBuilder sb = new StringBuilder();
    
    t = Integer.parseInt(br.readLine());

    for (int i = 0; i < t; i++) {
      StringTokenizer st = new StringTokenizer(br.readLine());
      n = Integer.parseInt(st.nextToken());
      k = Integer.parseInt(st.nextToken());

      st = new StringTokenizer(br.readLine());
      time = new int[n+1];
      for (int j = 1; j <= n; j++) {
       time[j] = Integer.parseInt(st.nextToken());
      }
      
      graph = new List[n + 1];
      for (int j = 0; j <= n; j++) {
        graph[j] = new ArrayList<>();  
      }

      inDegree = new int[n+1];
      for (int j = 0; j < k; j++) {
        st = new StringTokenizer(br.readLine());
        int a = Integer.parseInt(st.nextToken());
        int b = Integer.parseInt(st.nextToken());

        graph[a].add(b);
        inDegree[b]++;
      }

      w = Integer.parseInt(br.readLine());

      Queue<Integer> queue = new LinkedList<>();
      dp = new int[n + 1];
      for (int j = 1; j <= n; j++) {
        if (inDegree[j] == 0) {
          queue.add(j);
          dp[j] = time[j];
        }
      }

      while (!queue.isEmpty()) {
        int build = queue.poll();

        if (build == w) {
          sb.append(dp[w]).append("\n");
          break;
        }
        
        for (int next : graph[build]) {
          inDegree[next] -= 1;

          dp[next] = Math.max(dp[next], dp[build] + time[next]);
          if (inDegree[next] == 0) {
            queue.add(next);
          }
        }
      }
    }

    System.out.println(sb);
  }
}